File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.scala
import scala.collection.mutable

object Solution {
    def lengthOfLongestSubstring(s: String): Int = {


        var mapping : mutable.HashMap[Char, Int]= mutable.HashMap()
        var start = 0
        var len = 0
        if(s.isEmpty || s.length==1){
        return s.length
        }
        for((char, index)<-s.zipWithIndex){
        
        if(mapping.contains(char) && (mapping(char)>=start)){
            start = mapping(char)+1
            
        }
        mapping(char) = index
        len = math.max(len, (index-start)+1)
        
        }
        len

  }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        if not s:
            return 0
        longest_substr = 0

        sp = 0
        fp = 0
        mapping = {}
        #print(len(s))
        while(fp<len(s)):
            if s[fp] in mapping.keys() and mapping[s[fp]]>=sp:       
                sp = mapping[s[fp]]+1                
            else:
                longest_substr = max(longest_substr, (fp-sp)+1)
            #print(f"{sp=} {fp=} {mapping=} {longest_substr}")
            mapping[s[fp]] = fp    
            fp+=1

        return longest_substr

